### PR TITLE
Speed improvements

### PR DESCRIPTION
**Pull request by nallath**
I found some performance issues with the way that the cache was handled. 

On my system, this improved the speed of the handle_response function by ~25%.

### CODE REVIEW

**Comment by LipuFei** [comment]
@nallath For some reason, I can review and approve this PR...

**Comment by jstasiak** [comment]
Hey @nallath, thanks for taking a shot at this!

Can you update commit messages of every commit that's meant to speed things up with a reproducible way/code to measure the speed improvement and the results that you got (timings, number of iterations etc.)? This is crucial in order to review this set of patches.

**Comment by nallath** [comment]
Cura (www.github.com/ultimaker/Cura) is using this library, so what I did for each of these improvements is run specific functions that the profiler marked as being bottlenecks with kernprof (I know, it's not really a representative test, but it's the most punishing one I can come up with, since we have 100+ devices on this network spamming updates) 

This gives a fair bit of data, which I could put into the commit messages if that's something you want. I don't really have reproduction steps for it though (other than "Run this with a profiler").

**Comment by jstasiak** [comment]
I think this information + profile results before and after will be good enough. :)

**Comment by nallath** [comment]
Before:
Function: entries at line 1137

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
  1137                                               @profile
  1138                                               def entries(self):
  1139                                                   """Returns a list of all entries"""
  1140      1472       3135.0      2.1      0.5          if not self.cache:
  1141         2          3.0      1.5      0.0              return []
  1142                                                   else:
  1143                                                       # avoid size change during iteration by copying the cache
  1144      1470      28635.0     19.5      4.6              values = list(self.cache.values())
  1145      1470     596112.0    405.5     94.9              return reduce(lambda a, b: a + b, values)

after:
Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
  1140                                               @profile
  1141                                               def entries(self):
  1142                                                   """Returns a list of all entries"""
  1143      1806       2246.0      1.2      3.8          if not self._cache:
  1144         2          4.0      2.0      0.0              return []
  1145                                                   else:
  1146                                                       # avoid size change during iteration by copying the cache
  1147      1804      10055.0      5.6     17.1              values = list(self._cache.values())
  1148      1804      46591.0     25.8     79.1              return list(itertools.chain.from_iterable(values))

Previously it would take 405.5 per hit, with the changes it's 25.8 per hit

File: zeroconf/__init__.py [hunk 1]
```diff
@@ -1090,17 +1092,17 @@ class DNSCache:
     """A cache of DNS entries"""
 
     def __init__(self):
-        self.cache = {}  # type: Dict[str, List[DNSEntry]]
+        self._cache = {}  # type: Dict[str, List[DNSEntry]]
```
Comment (jstasiak) [inline zeroconf/__init__.py]: This change seems out of scope of this pull request – is it necessary for some reason?

**Comment by jstasiak** [comment]
> _Every_ commit that's meant to improve speed (I count 4 of those here) needs to contain a properly indented before/after profile

Before/after changes from that particular commit are applied, that is.

File: zeroconf/__init__.py (hunk 1 above)
Comment (nallath) [reply zeroconf/__init__.py]: It's not necessary, but I'm used to the boyscout rule (eg; leave the code you touched better than you left it).
In this case, the DNS cache uses a dict to do it's internal work, but it's not meant as an interface. As such, it should be protected / private. 
Anyway, I could also do that in a seperate pull request.

Comment (jstasiak) [reply zeroconf/__init__.py]: That's fair, although it's not clear that it's actually better this way (it's questionable and I'd not bother with it) and I don't think this principle applies to programming and pull requests, as unrelated improvements create noise.

Comment (nallath) [reply zeroconf/__init__.py]: I'll remove it from this pull request. We can have a more in depth discussion as to why this is better in the other pull request. I do have a bunch of arguments as to why this is better.

File: zeroconf/__init__.py [hunk 2]
```diff
@@ -483,15 +485,15 @@ def get_expiration_time(self, percent: float) -> float:
...
+        return max(0, (self._expiration_time - now) / 1000.0)
 
     def is_expired(self, now: float) -> bool:
         """Returns true if this record has expired."""
-        return self.get_expiration_time(100) <= now
+        return self._expiration_time <= now
 
     def is_stale(self, now):
         """Returns true if this record is at least half way expired."""
-        return self.get_expiration_time(50) <= now
```
Comment (Ghostkeeper) [inline zeroconf/__init__.py]: To get some statistics on this change, I did an experiment. I have a network with 170 devices running Zeroconf. Then we were listening for Zeroconf devices on that network. To get a more realistic test, the Zeroconf ServiceBrowser is ran on a separate thread from a main thread. On the main thread an I/O limited call to QNetworkManager is made every 2 seconds in order to include performance penalties due to thread switching. The experiment was ran on a MacBook Air 2015 (Intel Core i7-5650U) through Ubuntu 19.10 and Python 3.7. I left this running for exactly one hour, both before and after this commit.

Before this commit, there were 132107499 (132M) calls to the `get_expiration_time` function, totalling 141.647s (just over 2 minutes).

After this commit, there were 1661203 (1.6M) calls to the `get_expiration_time` function, totalling 2.068s. This saved about 2 minutes of processing time out of the total 60 minutes, on average 3.88% processing power on my CPU.

File: zeroconf/__init__.py [hunk 3]
```diff
@@ -2359,9 +2359,8 @@ def handle_response(self, msg: DNSIncoming) -> None:
         are held in the cache, and listeners are notified."""
         now = current_time_millis()
         for record in msg.answers:
-            expired = record.is_expired(now)
             if record in self.cache.entries():
-                if expired:
```
Comment (Ghostkeeper) [inline zeroconf/__init__.py]: I ran an experiment to test the performance of this change. I found no noticeable changes in performance due to this commit.

On a network with 170 devices, I ran a performance test. Every minute a request for broadcast is made. The Zeroconf ServiceBrowser is made to listen for incoming broadcasts. The ServiceBrowser was running on a separate thread while on the main thread an I/O-limited QNetworkRequest was made to include the effect of thread switching into my measurements. The experiment was ran on a MacBook Air 2015 (Intel Core i7-5650U) through Ubuntu 19.10 and Python 3.7. I left this running for exactly one hour, both before and after this single commit.

Before this commit, 139025204 (139M) calls were made to `DNSPointer.is_expired`, totalling 381.993s (just over 6 minutes).

After this commit, 140926403 (140M) calls were made to `DNSPointer.is_expired`, totalling 377.424s (just over 6 minutes).

I reckon that this falls within my measurement inaccuracy, so there is no noticeable change due to this commit in my experiment. Apparently `record` is in `self.cache.entries()` pretty much every time in my case.

However even though there was no performance benefit in practice, there is the possibility that there is a performance benefit for other cases. One could construe a test that resets the DNS cache every time to effectively measure start-up time for an application. I didn't take the effort to do this. In theory, it's easy to see that this change can only have benefits and no disadvantages. It's strictly doing less work, so I think this commit is still valid.

File: zeroconf/__init__.py [hunk 4]
```diff
@@ -651,15 +659,16 @@ def write(self, out):
...
     def __eq__(self, other):
         """Tests equality on priority, weight, port and server"""
         return (
             isinstance(other, DNSService)
-            and DNSEntry.__eq__(self, other)
             and self.priority == other.priority
             and self.weight == other.weight
             and self.port == other.port
             and self.server == other.server
+            and DNSEntry.__eq__(self, other)
```
Comment (Ghostkeeper) [inline zeroconf/__init__.py]: I ran an experiment to test the effectiveness of this change, measuring performance in a realistic situation.

I ran a performance test on a network with 170 devices running Zeroconf. On my computer there was a ServiceBrowser running on a separate thread while a timer ran on the main thread that forced a thread switch every 2 seconds (to include the effect of thread switching in the measurements). Every minute, a Zeroconf broadcast was made on the network. I ran this for an hour on a Macbook Air from 2015 (Intel Core i7-5650U) using Ubuntu 19.10 and Python 3.7, both before this commit and after.

These are the results of the performance tests:

| Function | Before count | Before time | Before time per count | After count | After time | After time per count | Time reduction |
|----|----|----|----|----|----|----|----|
|`DNSEntry.__eq__`|528|0.001s|1.9μs|538|0.001s|1.9μs|1.9%|
|`DNSPointer.__eq__`|24369256 (24.3M)|134.641s|5.5μs|25989573 (26.0M)|86.405s|3.3μs|39.8%|
|`DNSText.__eq__`|52966716 (53.0M)|190.640s|3.6μs|53604915 (53.6M)|169.104s|3.2μs|12.4%|
|`DNSService.__eq__`|52620538 (52.6M)|171.660s|3.3μs|56557448 (56.6M)|170.222s|3.0μs|7.8%|

So my conclusion from this are these:
* For `DNSEntry` my test was not conclusive. The sample size was not big enough.
* For `DNSPointer` and `DNSText` the performance was significantly better.
* For `DNSService` the performance was slightly better, on the edge of being statistically significant I'd say.

File: zeroconf/__init__.py [hunk 5]
```diff
@@ -1142,8 +1143,7 @@ def entries(self):
         else:
             # avoid size change during iteration by copying the cache
             values = list(self.cache.values())
-            return reduce(lambda a, b: a + b, values)
-
+            return list(itertools.chain.from_iterable(values))
```
Comment (Ghostkeeper) [inline zeroconf/__init__.py]: I didn't manage to figure out how to hit this in a profiled test. However Nallath himself seems to have profiler information on this change.

**Comment by jstasiak** [comment]
I'd be happy to just merge and release https://github.com/jstasiak/python-zeroconf/pull/205/commits/5b8149b71ddf299ead672b0dd9fb31aff16fc3b0 and https://github.com/jstasiak/python-zeroconf/pull/205/commits/f018cc0284d84b904953d8c9ccf6c87019c72955 as they seem to be the most important patches here and since @Ghostkeeper provided nice numbers I could copy them in the commit messages for historical visibility. Then we could have more time to actually review the other two patches. What do you say @nallath?

**Comment by nallath** [comment]
e2792f56d658584a9ce8155b607de9b96364e187 also provides a rather significant boost, of which I can get some stats in. 

I will put some effort into cleaning this up and putting the data in the commit messages today (I expect to end up with 3 improvements)

**Comment by nallath** [comment]
@jstasiak I've cleaned up the history and added the profiler information (In part, I've re-used the data provided by @Ghostkeeper)

**Comment by jstasiak** [comment]
Cheers. I just merged 3 out of 4 patches. I let myself reformat the commit messages to fit the usual line length and squash formatting fixes with the commits that introduced formatting breakage. Commits merged as:

* 157fc2003318 ("Significantly improve the speed of the entries function of the cache")
* 2e9699c542f6 ("Dont recalculate the expiration and stale time every update")
* 815ac77e9146 ("Change order of equality check to favor cheaper checks first")

I couldn't merge "Only check if an entry is expired when we need it" as it directly conflicts with a bugfix I was about to merge from a different pull request (1d39b3edd141093f9e579ab83377fe8f5ecb357d). If there's some optimization still to be had here feel free to open another pull request.

Thank you for the great speedup!

**Comment by jstasiak** [comment]
Version 0.24.1 with these changes just has been released.
